Code cleanup from query profile project

paul-rogers opened the pull request:
### Description

Code cleanup from query profile project. Cleanup changes done as separate PR to simplify review of the "substantial" changes.

* Fix spelling errors
* Fix Javadoc formatting
* Abstract out repeated test code
* Reuse constants in place of some string literals
* Fix up some parameterized types
* Reduce warnings reported by Eclipse
* Add jenv file and generated "pmd" files to .gitignore

There are no functional changes.

This PR has:

- [X] been self-reviewed.
- [X] been tested in a by running a full local Druid build & test cycle.

FrankChen021 commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -76,6 +76,7 @@
   private final boolean hasTimeout;
   private final long timeoutAtNanos;
+  @SuppressWarnings("unused")
   private final int queryPriority; // not currently used :(
```
Can we delete this unused variable?

FrankChen021 commented:
overall LGTM. CI has been re-triggered.

jihoonson commented on core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java:
```diff
@@ -312,14 +314,15 @@ private MergeCombinePartitioningAction(
       this.cancellationGizmo = cancellationGizmo;
     }
 
+    @SuppressWarnings("unchecked")
```
nit: I was looking for where the unchecked operation is used and realized that it is where `TERMINAL` is added to a queue. Could it be better to add this suppression at the line where unchecked operation is actually OK such as adding `TERMINAL` to a queue? The current change seems fine for now but I suppose this warning can do a legit thing if someone modifies some code in this method to use an unchecked operation. This comment is not a blocker.

jihoonson commented on processing/src/main/java/org/apache/druid/query/QueryMetrics.java:
```diff
@@ -32,47 +32,48 @@
- * ---------------------
- *  1. Skipping or partial filtering of particular dimensions or metrics entirely. Implementation could leave the body
- *  of the corresponding method empty, or implement random filtering like:
+ * <h4>Goals of QueryMetrics</h4>
```
nit: we don't publish javadoc, and thus prefer a format that is easily readable in IDEs than in web browsers.

jihoonson commented on processing/src/main/java/org/apache/druid/query/QueryMetrics.java:
```diff
@@ -262,36 +265,58 @@
   /**
    * Registers "query time" metric.
+   *
+   * Measures the time between a server thread starting to handle a query, and the response being fully written to
+   * the response output stream. Does not include time spent waiting in the server queue to acquire a thread.
```
Could it be clearer if you say the server thread is a Jetty server thread?

jihoonson commented on processing/src/main/java/org/apache/druid/query/QueryMetrics.java:
```diff
@@ -262,36 +265,58 @@
   /**
    * Registers "wait time" metric.
+   *
+   * Measures the total time segment-processing runnables spend waiting for execution in the processing thread pool.
```
```suggestion
   * Measures the total time segment-processing runnables spent waiting for execution in the processing thread pool.
```

jihoonson commented on processing/src/main/java/org/apache/druid/query/QueryMetrics.java:
```diff
@@ -262,36 +265,58 @@
   /**
    * Registers "cpu time" metric.
+   *
+   * Measures the total amount of CPU time spent by all threads (server and processing) involved in a query.
```
Will it be clearer if you mention that this metric includes segment processing wait time, segment processing time, and per-segment result merge time? I guess `server` seems a bit ambiguous to me since there are so many severs in druid.

jihoonson commented on processing/src/main/java/org/apache/druid/segment/data/Indexed.java:
```diff
@@ -26,6 +26,13 @@
 import java.util.Comparator;
 
+/**
+ * Indexed is a fixed-size, immutable, indexed set of values which allows
```
:+1: thanks for adding javadoc!

jihoonson reviewed (commented):
@paul-rogers thanks for adding javadocs. I left some comments about them.

paul-rogers commented:
@FrankChen021 and @jihoonson, thanks for the review! Addressed your comments. In some cases, it meant reverting the change (this is just a cleanup PR, and leaving things be is sometimes the simplest solution.)

paul-rogers commented:
Looks like some conflicts have appeared. Please take a look at the most recent changes. After that, I'll squash commits and resolve conflicts. (It is a pain to resolve conflicts with unsquashed commits.)

jihoonson reviewed (approved):
@paul-rogers thank you for your patience! The latest changes LGTM. +1 after conflicts are resolved.

paul-rogers commented:
@jihoonson, thanks much for the review. Resolved merge conflicts. Easiest way was to squash commits and rebase on master. Only the files mentioned in the conflict message above were changed. Did not do a full test run; let's see what the Travis run here tells us.

paul-rogers commented:
Looks like the build is failing due to incomplete tests. In file `ParallelMergeCombiningSequence.java`, we replaced this line:

```java
  out.offer(ResultBatch.TERMINAL);
```

With this one:

```java
  out.offer((ParallelMergeCombiningSequence.ResultBatch<T>) ResultBatch.TERMINAL);
```

Which is clearly only a compile-time change (to remove a warning.) Yet, the build fails with this error:

```text
Diff coverage statistics:
------------------------------------------------------------------------------
|     lines      |    branches    |   functions    |   path
------------------------------------------------------------------------------
|   0% (0/1)     | 100% (0/0)     | 100% (0/0)     | org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java
------------------------------------------------------------------------------
...
ERROR: Insufficient line coverage of 0% (0/1). Required 50%.
```

This error prevents the Stage 2 tests from running.

It seems we've got a situation where we can't clean up warnings in code until we figure out how to fully unit test the code. While having tests is a *Good Thing*, it seems we've wrapped ourselves around the axel because we didn't require tests when the code was written, but we do require someone else to write the tests to clean up warnings.

Suggestions?

jihoonson commented:
> It seems we've got a situation where we can't clean up warnings in code until we figure out how to fully unit test the code. While having tests is a Good Thing, it seems we've wrapped ourselves around the axel because we didn't require tests when the code was written, but we do require someone else to write the tests to clean up warnings.

@paul-rogers thanks for bringing this up. People, including myself, have been struggling with a similar issue from time to time. The coverage bot is good to have because it reduces the burden of code review. The reviewer should manually check test coverage otherwise. For your point, I don't think we can enforce the original author to add all tests that cover 100% of the change because it will be another way to wrap ourselves around the axel :slightly_smiling_face: Making a PR will become extremely hard if we raise the bar of coverage check to 100%. I think our current problem is that the bot is not smart enough to tell what are real changes that impact production code path and what are not. We should work on this to make the bot better over time.

For this particular PR, I think there could be 3 options for us.

- Not fixing the warning (or just suppressing the warning). The warning seems trivial and I don't mind living with it.
- Ignoring the test coverage check failure. This makes sense only when all tests in the phase 2 pass. Travis allows us to manually start those tests, but they should be started one by one by clicking the restart button in Travis. This is doable but there are more than 60 tests in the phase 2 and my wrist will hurt after restarting all of them :slightly_smiling_face: 
- Adding a new test. I prefer this option most as it's good to have a larger test coverage. It seems not that hard to add a test for this issue if you add a custom fjp pool that always explodes in `execute()` since it is always called in `MergeCombinePartitioningAction.compute()`. You may want to make `MergeCombinePartitioningAction` `VisibleForTesting` for easy testing.

jihoonson commented:
@paul-rogers thanks for updating the PR. I just retriggered the failed test which seems flaky. I will merge this PR after it passes.

jihoonson commented:
Merging this PR. Thanks @paul-rogers!

paul-rogers commented:
@jihoonson, thanks much for your help!
